From 477d7786641ce082a9308d7cfa72cfa1435a88f4 Mon Sep 17 00:00:00 2001
From: Andrew Kelley <andrew@ziglang.org>
Date: Wed, 10 Sep 2025 02:03:37 -0700
Subject: [PATCH 111/133] Sema: add missed logic to checkPtrAttributes
X-Developer-Signature: v=1; a=openpgp-sha256; l=2887; i=xtex@astrafall.org;
 h=from:subject; bh=bmWdzizRladKqtMQEyZNcuaBQ9Zz6qdpMtvUijvxPdg=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78ni0o9dbJSkrODCzPXz0l8dITcSgY9rhDXsrv+3Z0
 mRtm8bcMZGFQYyLwVJMkaXIsMGbVSedX3RZuSzMHFYmkCHSIg0MDAwMLAx8uYl5pUY6Rnqm2oZ6
 hoY6RjpGDFycAjDVFV8ZGT6JvXB4M++048nOz/mCnJNsLHJS13i0HPjY+1RwfVDM0r2MDNOnbbj
 lU2PWtG8y48wKza74/m+Lo3i6rJuyRWY4fnwXzwQA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

It wasn't checking bit pointer data.

(cherry picked from commit 725dec6aa7a7ec48625d16e5ea4e2c730f73d9dd)
---
 src/Sema.zig                                        | 13 +++++++++++++
 .../implicitly_increasing_pointer_alignment.zig     |  5 +++--
 .../implicitly_increasing_slice_alignment.zig       |  5 +++--
 3 files changed, 19 insertions(+), 4 deletions(-)

diff --git a/src/Sema.zig b/src/Sema.zig
index d0d8c9b26b3f..b0ad8069246c 100644
--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -30672,6 +30672,19 @@ fn checkPtrAttributes(sema: *Sema, dest_ty: Type, inst_ty: Type, in_memory_resul
         } };
         return false;
     }
+
+    if (inst_info.packed_offset.host_size != dest_info.packed_offset.host_size or
+        inst_info.packed_offset.bit_offset != dest_info.packed_offset.bit_offset)
+    {
+        in_memory_result.* = .{ .ptr_bit_range = .{
+            .actual_host = inst_info.packed_offset.host_size,
+            .wanted_host = dest_info.packed_offset.host_size,
+            .actual_offset = inst_info.packed_offset.bit_offset,
+            .wanted_offset = dest_info.packed_offset.bit_offset,
+        } };
+        return false;
+    }
+
     return true;
 }
 
diff --git a/test/cases/compile_errors/implicitly_increasing_pointer_alignment.zig b/test/cases/compile_errors/implicitly_increasing_pointer_alignment.zig
index fcecfa861117..47d88dfdf884 100644
--- a/test/cases/compile_errors/implicitly_increasing_pointer_alignment.zig
+++ b/test/cases/compile_errors/implicitly_increasing_pointer_alignment.zig
@@ -16,6 +16,7 @@ fn bar(x: *u32) void {
 // backend=stage2
 // target=native
 //
-// :8:9: error: expected type '*u32', found '*align(1) u32'
-// :8:9: note: pointer alignment '1' cannot cast into pointer alignment '4'
+// :8:9: error: expected type '*u32', found '*align(8:8:8) u32'
+// :8:9: note: pointer host size '8' cannot cast into pointer host size '0'
+// :8:9: note: pointer bit offset '8' cannot cast into pointer bit offset '0'
 // :11:11: note: parameter type declared here
diff --git a/test/cases/compile_errors/implicitly_increasing_slice_alignment.zig b/test/cases/compile_errors/implicitly_increasing_slice_alignment.zig
index 171936097dc9..18ce4262eb09 100644
--- a/test/cases/compile_errors/implicitly_increasing_slice_alignment.zig
+++ b/test/cases/compile_errors/implicitly_increasing_slice_alignment.zig
@@ -17,5 +17,6 @@ fn bar(x: []u32) void {
 // backend=stage2
 // target=native
 //
-// :9:22: error: expected type '*[1]u32', found '*align(1) u32'
-// :9:22: note: pointer alignment '1' cannot cast into pointer alignment '4'
+// :9:22: error: expected type '*[1]u32', found '*align(8:8:8) u32'
+// :9:22: note: pointer host size '8' cannot cast into pointer host size '0'
+// :9:22: note: pointer bit offset '8' cannot cast into pointer bit offset '0'
-- 
2.52.0

